src/utils/davix_cs3_utils.cpp

```diff
@@ -0,0 +1,116 @@
+#include <utils/davix_cs3_utils.hpp>
+#include <stdlib.h>
+
+namespace Davix{
+namespace reva {
+
+//
+// Credential 
+//
+
+// Constructor
+Credentials::Credentials() {
+  credMap = new CredentialMap();
+}
+
+// Destructor
+Credentials::~Credentials() {
+  delete credMap;
+  credMap = NULL;
+}
+
+// Copy constructor
+Credentials::Credentials(const Credentials& other) {
+  credMap = new CredentialMap(*other.credMap);
+}
+
+// Move constructor
+Credentials::Credentials(Credentials&& other) {
+  credMap = other.credMap;
+  other.credMap = new CredentialMap();
+}
+
+// Copy assignment operator
+Credentials& Credentials::operator=(const Credentials& other) {
+  credMap = new CredentialMap(*other.credMap);
+  return *this;
+}
+
+// Move assignment operator
+Credentials& Credentials::operator=(Credentials&& other) {
+  credMap = other.credMap;
+  other.credMap = new CredentialMap();
+  return *this;
+}
+
+bool Credentials::isEmpty() const {
+  return credMap->empty();
+}
+
+RevaToken Credentials::getToken(std::string uri) const{
+    std::string tkn = "";
+    if(credMap->find(uri) != credMap->end()){
+        tkn = (*credMap)[uri].token;
+    } 
+    return tkn;
+}
+
+void Credentials::getCredentialMap(CredentialMap & cmap) const {
+     for (CredentialMap::iterator itr = credMap->begin(); itr != credMap->end(); ++itr) {
+       Credential c {itr->second.token, itr->second.isDst};
+       cmap.insert(std::pair<std::string, Credential>(itr->first,c));
+     }
+}
+
+void Credentials::addCredentials(std::string uri, std::string token, bool token_write_access){
```

Comment:
`token` can be mode `const&` reference.

Reply:
Regarding `token_write_access`, I found it a bit difficult to convert from this role to `isDst` role, which is the flag's corresponding purpose in the `Credential` object.

Reply:
Earlier we misunderstood the implications of `token_write_access`. We were trying to add an extra var. I missed to correct it here. Changed it back to `token_write_access`.